### PR TITLE
Kevin boucher

### PR DESCRIPTION
**Pull request by kboucher**
Pull request for Kevin Boucher's solution to the Smith & Carson hiring code exercise.

**Demo**
https://user-images.githubusercontent.com/1943568/225469541-5f11aae8-d5ba-473c-aa37-d55e0efa33cc.mov

* Upgrade Ember.js version to 3.28.4 (not upgrading to v4 in case IE 11 support is still required)
* Run linters with `--fix` option (would like to change some of these rules)
* Add SASS support via `ember-cli-sass` addon
* Add `ember-bootstrap` addon
* Upgrade `ember-auto-import` to support `ember-bootstrap` addon
* Update router to make `candidates` the default route
* Import Smith & Carson fonts from Google fonts API
* Add base styles and layout elements and hide _Candidates_ link
* Add `ember-intl` addon for translations, update raw text in templates, convert candidates list to table and style _Add new_ button
* Update prettier tab size config (and apply)
* Add `ember-truth-helpers` addon
* Add Candidate route and form component
    * Add `CandidatesCandidateRoute`
    * Add `CandidatesCandidateController`
    * Add `ApplicantFormComponent`
    * Add Bootstrap `form-control` style overrides for invalid field treatment
    * Add `ember-test-selectors` addon
    * Remove `CandidatesCandidateController#addNew` action in favor of `LinkTo` detail route
    * Remove `outlet` from list view template
* Ensure candidates list is refreshed after adding/updating a candidate
* Update README.md structure
* Add whitespace to the bottom of the main layout container
* Add POST route to Phalcon server index
* Standardize invalid form field treatment (leverage Bootstrap `is-invalid`)

### CODE REVIEW

File: README.md
Comment (kboucher) [inline README.md:1]: Just tweaked the markdown a bit here.

File: client/.eslintrc.js
Comment (kboucher) [inline client/.eslintrc.js:1]: Spacing changes after update of prettier config.

File: client/.eslintignore [hunk 1]
```diff
@@ -13,6 +13,7 @@
 # misc
 /coverage/
 !.*
+.*/
```
Comment (kboucher) [inline client/.eslintignore:16]: `ember-cli-update` changes.

File: client/.template-lintrc.js [hunk 2]
```diff
@@ -1,5 +1,5 @@
 'use strict';
 
 module.exports = {
-  extends: 'octane',
+    extends: 'recommended',
```
Comment (kboucher) [inline client/.template-lintrc.js:4]: Spacing changes after update of prettier config.

File: client/.travis.yml [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ---
 language: node_js
 node_js:
-  - "10"
+  - "12"
```
Comment (kboucher) [inline client/.travis.yml:4]: `ember-cli-update` change.

File: client/app/adapters/application.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 import JSONAPIAdapter from '@ember-data/adapter/json-api';
 
 export default class ApplicationAdapter extends JSONAPIAdapter {
-  namespace = 'api';
+    namespace = 'api';
```
Comment (kboucher) [inline client/app/adapters/application.js:4]: Spacing changes after update of prettier config.

File: client/app/app.js [hunk 5]
```diff
@@ -4,9 +4,9 @@ import loadInitializers from 'ember-load-initializers';
 import config from 'client/config/environment';
 
 export default class App extends Application {
-  modulePrefix = config.modulePrefix;
-  podModulePrefix = config.podModulePrefix;
-  Resolver = Resolver;
+    modulePrefix = config.modulePrefix;
+    podModulePrefix = config.podModulePrefix;
+    Resolver = Resolver;
```
Comment (kboucher) [inline client/app/app.js:9]: Spacing changes after update of prettier config.

File: client/app/components/applicant-form/template.hbs [hunk 6]
```diff
@@ -0,0 +1,79 @@
+<form
+    class="mt-3"
+    data-test-form="applicant"
+    id={{this.formId}}
+    {{on "submit" @onSubmit}}
+>
+    <div class="mb-3">
+        <label
+            class="form-label"
+            for={{this.nameInputId}}
```
Comment (kboucher) [inline client/app/components/applicant-form/template.hbs:10]: Would normally handle this with reusable Ember form components.

File: client/app/controllers/candidates.js [hunk 7]
```diff
@@ -1,10 +1,9 @@
...
 
-export default class CandidatesController extends Controller {
+/**
+    Provides Ember.Controller functionality for `CandidatesRoute`.
 
-  @action
-  addNew () {
-    alert('Not implemented')
-  }
-}
```
Comment (kboucher) [inline client/app/controllers/candidates.js:10]: _Add new_ button is now a simple `LinkTo` that transitions to the detail view.

File: client/app/formats.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+export default {
```
Comment (kboucher) [inline client/app/formats.js:1]: Added `ember-intl` localization addon.

File: client/app/index.html [hunk 9]
```diff
@@ -11,6 +11,10 @@
 
     <link integrity="" rel="stylesheet" href="{{rootURL}}assets/vendor.css">
     <link integrity="" rel="stylesheet" href="{{rootURL}}assets/client.css">
+    <link rel="preconnect" href="https://fonts.googleapis.com">
+    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
+    <link href="https://fonts.googleapis.com/css2?family=Source+Sans+Pro:ital,wght@0,200;0,400;0,700;1,200;1,400;1,700&display=swap" rel="stylesheet">
+    <link href="https://fonts.googleapis.com/css2?family=Playfair+Display:wght@600;800&family=Source+Sans+Pro:ital,wght@0,200;0,400;0,700;1,200;1,400;1,700&display=swap" rel="stylesheet">
```
Comment (kboucher) [inline client/app/index.html:17]: Import Smith & Carson fonts.

File: client/app/models/applicant.js [hunk 10]
```diff
@@ -1,6 +1,6 @@
-import Model, {attr} from '@ember-data/model';
+import Model, { attr } from '@ember-data/model';
 
 export default class ApplicantModel extends Model {
-  @attr('string') name
-  @attr('number') age
+    @attr('string') name;
+    @attr('number') age;
```
Comment (kboucher) [inline client/app/models/applicant.js:5]: Spacing changes after update of prettier config (and `lint --fix` update).

File: client/app/router.js [hunk 11]
```diff
@@ -2,10 +2,12 @@ import EmberRouter from '@ember/routing/router';
 import config from 'client/config/environment';
 
 export default class Router extends EmberRouter {
-  location = config.locationType;
-  rootURL = config.rootURL;
+    location = config.locationType;
+    rootURL = config.rootURL;
```
Comment (kboucher) [inline client/app/router.js:6]: Spacing changes after update of prettier config.

File: client/app/router.js [hunk 12]
```diff
@@ -2,10 +2,12 @@ import EmberRouter from '@ember/routing/router';
...
 }
 
 Router.map(function () {
-  this.route('candidates');
+    this.route('candidates', { path: '/' }, function () {
```
Comment (kboucher) [inline client/app/router.js:10]: Make `candidates` the default route.

File: client/app/router.js [hunk 13]
```diff
@@ -2,10 +2,12 @@ import EmberRouter from '@ember/routing/router';
...
+        this.route('candidate', { path: ':id' });
```
Comment (kboucher) [inline client/app/router.js:11]: Add child detail route.

File: client/app/routes/candidates.js [hunk 14]
```diff
@@ -1,11 +1,30 @@
 import Route from '@ember/routing/route';
+import { action } from '@ember/object';
 import { inject as service } from '@ember/service';
 
+/**
+    Provides `Ember.Route` functionality for Candidates list view.
+
+    @module CandidatesRoute
+    @extends Ember.Route
+ */
```
Comment (kboucher) [inline client/app/routes/candidates.js:10]: Add jsDocs.

File: client/app/routes/candidates.js [hunk 15]
```diff
@@ -1,11 +1,30 @@
...
+
+    /**
+        Returns list of `ApplicantRecord`s from server
+
+        @method model
+        @returns {Array} Collection of `ApplicantRecord`s
+     */
+    model() {
+        return this.store.query('applicant', {});
+    }
```
Comment (kboucher) [inline client/app/routes/candidates.js:22]: Spacing changes after update of prettier config.

File: client/app/routes/candidates.js [hunk 16]
```diff
@@ -1,11 +1,30 @@
...
-  model ()
-  {
-    return this.store.query('applicant', {})
-  }
+    /**
+        @method refreshModel
+     */
+    @action refreshModel() {
+        this.refresh();
+    }
```
Comment (kboucher) [inline client/app/routes/candidates.js:29]: Add `refreshModel` action called from child detail route.

File: client/app/styles/app.css
Comment (kboucher) [inline client/app/styles/app.css:1]: Converted to SASS (`.scss` file).

File: client/app/styles/bootstrap-overrides/_forms.scss [hunk 17]
```diff
@@ -0,0 +1,3 @@
+input.form-control.touched:invalid {
+    @extend .is-invalid;
+}
```
Comment (kboucher) [inline client/app/styles/bootstrap-overrides/_forms.scss:3]: Invalid form field treatment.

File: client/app/styles/bootstrap-overrides/_navbar.scss [hunk 18]
```diff
@@ -0,0 +1,6 @@
+.navbar {
+    .navbar-brand {
+        font-family: "Playfair Display", serif;
```
Comment (kboucher) [inline client/app/styles/bootstrap-overrides/_navbar.scss:3]: S&C display font.

File: client/app/styles/layouts/_application.scss [hunk 19]
```diff
@@ -0,0 +1,8 @@
+html {
+    font-size: 100%;
```
Comment (kboucher) [inline client/app/styles/layouts/_application.scss:2]: Ensure font sizes are responsive and respect the user's client configuration.

File: client/app/styles/layouts/_application.scss [hunk 20]
```diff
@@ -0,0 +1,8 @@
...
+}
+
+body {
+    font-family: "Source Sans Pro", sans-serif;
```
Comment (kboucher) [inline client/app/styles/layouts/_application.scss:6]: S&C content font.

File: client/app/templates/application.hbs [hunk 21]
```diff
@@ -1,5 +1,19 @@
-{{page-title "Client"}}
+{{page-title (t "page_titles.candidates")}}
```
Comment (kboucher) [inline client/app/templates/application.hbs:1]: Leverage `ember-intl` to deliver all UI text as translation values (from `en-us.json`).

File: client/app/templates/application.hbs [hunk 22]
```diff
@@ -1,5 +1,19 @@
...
\ No newline at end of file
+{{!--
+    Candidates is now the default route
+
+    {{link-to "Candidates" "candidates"}}
+ --}}
+
+<div class="container pb-3">
+    {{outlet}}
+</div>
```
Comment (kboucher) [inline client/app/templates/application.hbs:19]: Add container around primary content.

File: client/app/templates/candidates.hbs [hunk 23]
```diff
@@ -1,15 +1,32 @@
...
+                <td>{{applicant.name}}</td>
+                <td>{{applicant.age}}</td>
+            </tr>
+        {{else}}
+            <tr>
+                <td colspan="2">{{t "status_messages.no_records_found"}}</td>
+            </tr>
+        {{/each}}
+    </tbody>
+</table>
```
Comment (kboucher) [inline client/app/templates/candidates.hbs:24]: Convert list to table.

File: client/app/templates/candidates/candidate.hbs [hunk 24]
```diff
@@ -0,0 +1,10 @@
+{{page-title (t "page_titles.candidate")}}
+
+<h1>{{t "page_titles.candidate"}}</h1>
+
+<ApplicantForm
+    @applicantAge={{this.model.age}}
+    @applicantName={{this.model.name}}
```
Comment (kboucher) [inline client/app/templates/candidates/candidate.hbs:7]: Passing these individually as `args` are watched by default. Just passing the `applicant` would require jumping through a few more hoops to activate observers.

File: client/app/templates/candidates/candidate.hbs [hunk 25]
```diff
@@ -0,0 +1,10 @@
...
+    @onCancel={{this.onCancel}}
+    @onSubmit={{this.onSubmit}}
```
Comment (kboucher) [inline client/app/templates/candidates/candidate.hbs:9]: Handle model changes in the invoking context (DDAU).

File: client/config/ember-cli-update.json [hunk 26]
```diff
@@ -3,14 +3,13 @@
   "packages": [
     {
       "name": "ember-cli",
-      "version": "3.25.0",
+      "version": "3.28.4",
```
Comment (kboucher) [inline client/config/ember-cli-update.json:6]: Upgraded Ember to latest 3.* (did not go to 4.* to preserve IE 11 support).

File: client/config/ember-cli-update.json [hunk 27]
```diff
@@ -3,14 +3,13 @@
...
       "blueprints": [
         {
           "name": "app",
           "outputRepo": "https://github.com/ember-cli/ember-new-output",
           "codemodsSource": "ember-app-codemods-manifest@1",
-          "isBaseBlueprint": true,
-          "options": []
+          "isBaseBlueprint": true
```
Comment (kboucher) [inline client/config/ember-cli-update.json:12]: Ember upgrade mods.

File: client/config/ember-intl.js [hunk 28]
```diff
@@ -0,0 +1,95 @@
+/* jshint node:true */
```
Comment (kboucher) [inline client/config/ember-intl.js:1]: `ember-intl` config file.

File: client/config/environment.js [hunk 29]
```diff
@@ -1,51 +1,51 @@
 'use strict';
 
 module.exports = function (environment) {
```
Comment (kboucher) [inline client/config/environment.js:3]: Spacing changes after update of prettier config. ⬇️

File: client/package.json [hunk 30]
```diff
@@ -12,7 +12,7 @@
   },
   "scripts": {
     "build": "ember build --environment=production",
-    "lint": "npm-run-all --aggregate-output --continue-on-error --parallel 'lint:!(fix)'",
+    "lint": "npm-run-all --aggregate-output --continue-on-error --parallel \"lint:!(fix)\"",
```
Comment (kboucher) [inline client/package.json:15]: Must be an Ember. upgrade mod.

File: client/package.json [hunk 31]
```diff
@@ -24,44 +24,53 @@
   },
   "devDependencies": {
     "@ember/optional-features": "^2.0.0",
```
Comment (kboucher) [inline client/package.json:26]: Ember upgrade, add dependencies, etc. ⬇️

File: client/testem.js [hunk 32]
```diff
@@ -1,23 +1,23 @@
 'use strict';
 
 module.exports = {
```
Comment (kboucher) [inline client/testem.js:3]: Spacing changes after update of prettier config. ⬇️

File: client/tests/index.html [hunk 33]
```diff
@@ -28,7 +28,7 @@
       </div>
     </div>
 
-    <script src="/testem.js" integrity=""></script>
+    <script src="/testem.js" integrity="" data-embroider-ignore></script>
```
Comment (kboucher) [inline client/tests/index.html:31]: Ember upgrade mod.

File: client/tests/integration/components/applicant-form/component-test.js [hunk 34]
```diff
@@ -0,0 +1,32 @@
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import { render } from '@ember/test-helpers';
+import { hbs } from 'ember-cli-htmlbars';
+
+module('Integration | Component | applicant-form', function (hooks) {
+    setupRenderingTest(hooks);
+
```
Comment (kboucher) [inline client/tests/integration/components/applicant-form/component-test.js:8]: Very basic component integration test (leveraging `ember-test-selectors`).

File: client/tests/unit/adapters/application-test.js [hunk 35]
```diff
@@ -1,12 +1,12 @@
 import { module, test } from 'qunit';
 import { setupTest } from 'ember-qunit';
```
Comment (kboucher) [inline client/tests/unit/adapters/application-test.js:3]: Spacing changes after update of prettier config. ⬇️

File: client/translations/en-us.json [hunk 36]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (kboucher) [inline client/translations/en-us.json:1]: Translation values ⬇️

File: server/public/index.php [hunk 37]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
   }
 );
 
+$app->post(
+    '/api/applicants',
+    function () use ($app) {
+        $applicantPayload = $app->request->getJsonRawBody();
```
Comment (kboucher) [inline server/public/index.php:77]: Get the payload from the request.

File: server/public/index.php [hunk 38]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+        $applicantAttributes = $applicantPayload->data->attributes;
```
Comment (kboucher) [inline server/public/index.php:78]: Get the attributes of the JSON API payload.

File: server/public/index.php [hunk 39]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+
+        $name = $applicantAttributes->name;
+        $age = $applicantAttributes->age;
```
Comment (kboucher) [inline server/public/index.php:81]: Declare (and assign) `name`/`age` variables.

File: server/public/index.php [hunk 40]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+
+        $phql = "INSERT INTO App\Models\Candidates (name, age) VALUES (:name:, :age:)";
```
Comment (kboucher) [inline server/public/index.php:83]: Define `INSERT` statement.

File: server/public/index.php [hunk 41]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+
+        $status = $app
+          ->modelsManager
+          ->executeQuery(
+                $phql,
+                [
+                    'age' => $age,
+                    'name' => $name,
+                ]
+            );
```
Comment (kboucher) [inline server/public/index.php:93]: Execute `INSERT` statement (defining `age`/`name` segments).

File: server/public/index.php [hunk 42]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+
+        $response = new Response();
```
Comment (kboucher) [inline server/public/index.php:95]: Create a `Response`.

File: server/public/index.php [hunk 43]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+
+        if ($status->success() === true) {
```
Comment (kboucher) [inline server/public/index.php:97]: Set `Response` HTTP status code, header, and JSON body ⬇️ .

File: server/public/index.php [hunk 44]
```diff
@@ -71,4 +71,68 @@ function () use ($app) {
...
+                      'id' => $status->getModel()->id,
+                      'attributes' => [
+                        'name' => $name,
+                        'age' => $age,
+                      ]
+                  ],
+                ]
+            );
+        } else {
+            // Generic error handler
```
Comment (kboucher) [inline server/public/index.php:118]: Handle errors ⬇️

**Comment by marceloandrader** [comment]
Thanks

**Pull request closed without merging**